### PR TITLE
[USM] Postgres telemetry, add tag "state" to metric usm.postgres.query_length_bucket<N>.

### PR DESCRIPTION
**Pull request by yuri-lipnesh**
### What does this PR do?
Add tag "state:" for Postgres telemetry "usm.postgres.query_length_bucket_<N>", where N is 1...10

### Motivation

Postgres communication uses a message protocol to exchange data between the client and the server. The number of messages exchanged can vary depending on the request. System-probe fetches a limited number of messages and looks for attributes such as table name and operation (SELECT, DELETE, etc.).

System-probe processes dedicated telemetry to determine the number of messages exchanged and fills the buckets according to the number of messages. Currently, it processes 10 buckets with 15 messages per bucket.

With this code, each bucket is assigned one of following tags:
This tag has values:
  "state:table_and_op" - table name and operation was found in query
  "state:no_operation" - table operation was not found in query
  "state:no_table_name" - table name was. not found in query
  "state:no_table_no_op" - both table name and operation was not found in query

This tag allows to understand whether system-probe can discover the desired Postgres attributes despite the limited number of fetched Postgres messages.

### Additional Notes

The added tag is used to evaluate the accuracy of Postgres monitoring.

### CODE REVIEW

File: pkg/network/protocols/postgres/telemetry.go [hunk 1]
```diff
@@ -88,16 +88,20 @@ func (t *Telemetry) getBucketIndex(querySize int) int {
...
+	fulfilled := false
 	if eventWrapper.Operation() == UnknownOP {
 		t.failedOperationExtraction.Add(1)
+	} else {
+		fulfilled = true
 	}
 	if eventWrapper.TableName() == "UNKNOWN" {
 		t.failedTableNameExtraction.Add(1)
+	} else {
+		fulfilled = true
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:100]: As I see it, currently, `fulfilled` can be true in both cases. I think having the ability to understand what failed could give us an advantage. For example, if we were unable to extract the operation but were able to extract the table name, we wouldn't be able to identify those cases.

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/postgres/telemetry_test.go [hunk 2]
```diff
@@ -187,9 +187,17 @@ func createEbpfEvent(querySize int) *ebpf.EbpfEvent {
...
+		expectFulfilledCounter = true
+	}
 	for i := 0; i < len(tel.queryLengthBuckets); i++ {
-		assert.Equal(t, expected.queryLength[i], tel.queryLengthBuckets[i].Get(), "queryLength for bucket %d count is incorrect", i)
+		assert.Equal(t, expected.queryLength[i], tel.queryLengthBuckets[i].Get(expectFulfilledCounter), "queryLength for bucket %d count is incorrect", i)
 	}
-	assert.Equal(t, expected.failedTableNameExtraction, tel.failedTableNameExtraction.Get(), "failedTableNameExtraction count is incorrect")
-	assert.Equal(t, expected.failedOperationExtraction, tel.failedOperationExtraction.Get(), "failedOperationExtraction count is incorrect")
+	assert.Equal(t, expected.failedTableNameExtraction, countFailedTableName, "failedTableNameExtraction count is incorrect")
+	assert.Equal(t, expected.failedOperationExtraction, countFailedOperation, "failedOperationExtraction count is incorrect")
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry_test.go:214]: I adjusted the test using the newly defined tags.

File: pkg/network/protocols/postgres/telemetry.go [hunk 3]
```diff
@@ -22,12 +22,76 @@ const (
...
+	return &resultAwareCounter{
+		countTableAndOperation:  metricGroup.NewCounter(metricName, append(tags, "fulfilled:table_and_op")...),
+		countOperationNotFound:  metricGroup.NewCounter(metricName, append(tags, "fulfilled:no_operation")...),
+		countTableNameNotFound:  metricGroup.NewCounter(metricName, append(tags, "fulfilled:no_table_name")...),
+		countTableAndOpNotFound: metricGroup.NewCounter(metricName, append(tags, "fulfilled:no_table_no_op")...),
+	}
+}
+
+// add adds delta to the counter of completed results or counter incomplete results.
+func (c *resultAwareCounter) add(delta int64, fulfilled counterEnum) {
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:57]: rename:
delta -> value
fulfilled -> state

File: pkg/network/protocols/postgres/telemetry.go [hunk 4]
```diff
@@ -22,12 +22,76 @@ const (
...
+	case fulfillTableAndOperation:
+		c.countTableAndOperation.Add(delta)
+	case fulfillOperationNotFound:
+		c.countOperationNotFound.Add(delta)
+	case fulfillTableNameNotFound:
+		c.countTableNameNotFound.Add(delta)
+	case fulfillTableAndOpNotFound:
+		c.countTableAndOpNotFound.Add(delta)
+	default:
+		log.Errorf("Add bucket counter, undefined enum.")
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:68]:
```suggestion
		log.Errorf("Add bucket counter, undefined enum - %v", fulfilled)
```

File: pkg/network/protocols/postgres/telemetry.go [hunk 5]
```diff
@@ -22,12 +22,76 @@ const (
...
+	}
+}
+
+// get returns the counter value based on the result.
+func (c *resultAwareCounter) get(fulfilled counterEnum) int64 {
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:73]: same comments as replied on `add`

File: pkg/network/protocols/postgres/telemetry.go [hunk 6]
```diff
@@ -22,12 +22,76 @@ const (
...
+		return c.countTableAndOperation.Get()
+	case fulfillOperationNotFound:
+		return c.countOperationNotFound.Get()
+	case fulfillTableNameNotFound:
+		return c.countTableNameNotFound.Get()
+	case fulfillTableAndOpNotFound:
+		return c.countTableAndOpNotFound.Get()
+	default:
+		log.Errorf("Get bucket counter, undefined enum.")
+		return -1
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:85]: why -1?
use the default value `0`
generally, metrics are non negative objects, returning signed int64 does not make sense

File: pkg/network/protocols/postgres/telemetry.go [hunk 7]
```diff
@@ -88,16 +152,22 @@ func (t *Telemetry) getBucketIndex(querySize int) int {
 func (t *Telemetry) Count(tx *ebpf.EbpfEvent, eventWrapper *EventWrapper) {
 	querySize := int(tx.Tx.Original_query_size)
 
-	bucketIndex := t.getBucketIndex(querySize)
-	if bucketIndex >= 0 && bucketIndex < len(t.queryLengthBuckets) {
-		t.queryLengthBuckets[bucketIndex].Add(1)
-	}
-
+	var fulfilled counterEnum = fulfillTableAndOperation
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:155]:
```suggestion
	fulfilled := fulfillTableAndOperation
```

File: pkg/network/protocols/postgres/telemetry.go [hunk 8]
```diff
@@ -88,16 +152,22 @@ func (t *Telemetry) getBucketIndex(querySize int) int {
...
 		t.failedTableNameExtraction.Add(1)
+		if fulfilled == fulfillOperationNotFound {
+			fulfilled = fulfillTableAndOpNotFound
+		} else {
+			fulfilled = fulfillTableNameNotFound
+		}
+	}
+	bucketIndex := t.getBucketIndex(querySize)
+	if bucketIndex >= 0 && bucketIndex < len(t.queryLengthBuckets) {
+		t.queryLengthBuckets[bucketIndex].add(1, fulfilled)
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:170]: if the only usage of `add` is `add(1, fulfiled)` then let's change it to `inc(fulfilled)`

File: pkg/network/protocols/postgres/telemetry_test.go [hunk 9]
```diff
@@ -187,9 +187,29 @@ func createEbpfEvent(querySize int) *ebpf.EbpfEvent {
 }
 
 func verifyTelemetry(t *testing.T, tel *Telemetry, expected telemetryResults) {
+
+	var expectFulfill counterEnum = fulfillTableAndOperation
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry_test.go:191]:
```suggestion
	expectFulfill := fulfillTableAndOperation
```

**Review by amitslavin — CHANGES_REQUESTED** [review]

**Comment by amitslavin** [comment]
Please update the PR description to reflect your latest changes.

File: pkg/network/protocols/postgres/operations.go [hunk 10]
```diff
@@ -29,6 +29,12 @@ const (
 	AlterTableOP
 	// TruncateTableOP represents a TRUNCATE operation.
 	TruncateTableOP
+	// ShowOP represents the 'SHOW' command to display the current setting of run-time parameters
+	ShowOP
+	// BeginOP represents the 'BEGIN' command to initiate a transaction
+	BeginOP
+	// CommitOP represents the 'COMMIT' command commits the transaction
+	CommitOP
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/operations.go:37]: Why did you add this to the PR?

Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/operations.go:37]: These operations are used in Postgres queries, discovering these operations will reduce errors when operation was not found.

Comment (amitslavin) [reply pkg/network/protocols/postgres/operations.go:37]: This is outside the scope of this PR. We have Jira tickets to add support for these operations in the future. ([example](https://datadoghq.atlassian.net/browse/USMON-1185))

Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/operations.go:37]: Do you want me to do it first? this is source of undefined counters

File: pkg/network/protocols/postgres/telemetry.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: i removed this error

File: pkg/network/protocols/postgres/telemetry.go [hunk 11]
```diff
@@ -22,12 +22,76 @@ const (
...
+	fulfillTableAndOperation counterEnum = iota + 1
+	fulfillOperationNotFound
+	fulfillTableNameNotFound
+	fulfillTableAndOpNotFound
+)
+
+// resultAwareCounter stores counter when goal was achieved and counter when target not found.
+type resultAwareCounter struct {
+	// countTableAndOperation counts the number of successfully retrieved table name and operation.
+	countTableAndOperation *libtelemetry.Counter
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: I've changed name of struct

File: pkg/network/protocols/postgres/telemetry.go [hunk 12]
```diff
@@ -22,12 +22,76 @@ const (
 	numberOfBucketsSmallerThanMaxBufferSize = 3
 )
 
+type counterEnum int
+
+const (
+	fulfillTableAndOperation counterEnum = iota + 1
+	fulfillOperationNotFound
+	fulfillTableNameNotFound
+	fulfillTableAndOpNotFound
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: renamed

File: pkg/network/protocols/postgres/telemetry_test.go [hunk 13]
```diff
@@ -187,9 +187,29 @@ func createEbpfEvent(querySize int) *ebpf.EbpfEvent {
...
+		expectFulfill = fulfillTableAndOpNotFound
+	}
+
 	for i := 0; i < len(tel.queryLengthBuckets); i++ {
-		assert.Equal(t, expected.queryLength[i], tel.queryLengthBuckets[i].Get(), "queryLength for bucket %d count is incorrect", i)
+		assert.Equal(t, expected.queryLength[i], tel.queryLengthBuckets[i].get(expectFulfill), "queryLength for bucket %d count is incorrect", i)
 	}
-	assert.Equal(t, expected.failedTableNameExtraction, tel.failedTableNameExtraction.Get(), "failedTableNameExtraction count is incorrect")
-	assert.Equal(t, expected.failedOperationExtraction, tel.failedOperationExtraction.Get(), "failedOperationExtraction count is incorrect")
+	assert.Equal(t, expected.failedTableNameExtraction, countFailedTableName, "failedTableNameExtraction count is incorrect")
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry_test.go]: done

File: pkg/network/protocols/postgres/telemetry.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: done

File: pkg/network/protocols/postgres/telemetry.go (hunk 5 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: done

File: pkg/network/protocols/postgres/telemetry.go (hunk 7 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: changed

File: pkg/network/protocols/postgres/telemetry.go (hunk 6 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: I've changed it

File: pkg/network/protocols/postgres/telemetry.go (hunk 8 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: changed

File: pkg/network/protocols/postgres/telemetry_test.go (hunk 9 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry_test.go]: done

File: pkg/network/protocols/postgres/telemetry.go (hunk 8 above)
Comment (amitslavin) [reply pkg/network/protocols/postgres/telemetry.go]: Looks like it is still relevant

File: pkg/network/protocols/postgres/telemetry_test.go [hunk 14]
```diff
@@ -187,8 +194,9 @@ func createEbpfEvent(querySize int) *ebpf.EbpfEvent {
 }
 
 func verifyTelemetry(t *testing.T, tel *Telemetry, expected telemetryResults) {
+
 	for i := 0; i < len(tel.queryLengthBuckets); i++ {
-		assert.Equal(t, expected.queryLength[i], tel.queryLengthBuckets[i].Get(), "queryLength for bucket %d count is incorrect", i)
+		assert.Equal(t, expected.queryLength[i], tel.queryLengthBuckets[i].get(expected.counterState), "queryLength for bucket %d count is incorrect", i)
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry_test.go:199]: After the changes you introduced, the checks are performing different functionality. You are now validating that the metrics you added are correct in that assertion. You need to modify the error message in case of failure, as the current message could be misleading if the test fails.

File: pkg/network/protocols/postgres/telemetry.go [hunk 15]
```diff
@@ -22,18 +22,81 @@ const (
...
+	return &extractionFailureCounter{
+		countTableAndOperationFound: metricGroup.NewCounter(metricName, append(tags, "state:table_and_op")...),
+		countOperationNotFound:      metricGroup.NewCounter(metricName, append(tags, "state:no_operation")...),
+		countTableNameNotFound:      metricGroup.NewCounter(metricName, append(tags, "state:no_table_name")...),
+		countTableAndOpNotFound:     metricGroup.NewCounter(metricName, append(tags, "state:no_table_no_op")...),
+	}
+}
+
+// inc increments the counter of completed results or counter incomplete results.
+func (c *extractionFailureCounter) inc(value int64, state counterStateEnum) {
```
Comment (guyarb) [inline pkg/network/protocols/postgres/telemetry.go:57]: inc - is usually meant for increasing the counter by 1, no need to get value

File: pkg/network/protocols/postgres/telemetry.go [hunk 16]
```diff
@@ -22,18 +22,81 @@ const (
...
+	case tableAndOperation:
+		c.countTableAndOperationFound.Add(value)
+	case operationNotFound:
+		c.countOperationNotFound.Add(value)
+	case tableNameNotFound:
+		c.countTableNameNotFound.Add(value)
+	case tableAndOpNotFound:
+		c.countTableAndOpNotFound.Add(value)
+	default:
+		log.Errorf("Add bucket counter, undefined enum: %v", state)
```
Comment (guyarb) [inline pkg/network/protocols/postgres/telemetry.go:68]: logs should not start with capital letter

File: pkg/network/protocols/postgres/telemetry.go (hunk 4 above)
Comment (amitslavin) [reply pkg/network/protocols/postgres/telemetry.go]: I am still not sure what is the `Add bucket counter,` meaning

File: pkg/network/protocols/postgres/telemetry.go [hunk 17]
```diff
@@ -22,18 +22,96 @@ const (
...
+type counterStateEnum int
+
+const (
+	tableAndOperation counterStateEnum = iota + 1
+	operationNotFound
+	tableNameNotFound
+	tableAndOpNotFound
+)
+
+func (e counterStateEnum) String() string {
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:34]: Where do you use this function?

File: pkg/network/protocols/postgres/telemetry.go [hunk 18]
```diff
@@ -22,18 +22,96 @@ const (
...
+	case tableAndOperation:
+		c.countTableAndOperationFound.Add(1)
+	case operationNotFound:
+		c.countOperationNotFound.Add(1)
+	case tableNameNotFound:
+		c.countTableNameNotFound.Add(1)
+	case tableAndOpNotFound:
+		c.countTableAndOpNotFound.Add(1)
+	default:
+		log.Errorf("postgres query length counter increment function received undefined state: %v", state)
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:83]:
```suggestion
		log.Errorf("unable to increment extractionFailureCounter due to undefined state: %v", state)
```

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/postgres/telemetry_test.go (hunk 14 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry_test.go]: changed

File: pkg/network/protocols/postgres/telemetry.go (hunk 18 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go]: changed as suggested

Comment (amitslavin) [reply pkg/network/protocols/postgres/telemetry.go]: You have committed it wrong and now you have format issues

File: pkg/network/protocols/postgres/telemetry.go [hunk 19]
```diff
@@ -22,18 +22,96 @@ const (
...
+	case tableAndOperation:
+		c.countTableAndOperationFound.Add(1)
+	case operationNotFound:
+		c.countOperationNotFound.Add(1)
+	case tableNameNotFound:
+		c.countTableNameNotFound.Add(1)
+	case tableAndOpNotFound:
+		c.countTableAndOpNotFound.Add(1)
+	default:
+		log.Errorf("\"unable to increment extractionFailureCounter due to undefined state: %v\", state", state)
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:83]:
```suggestion
		log.Errorf("unable to increment extractionFailureCounter due to undefined state: %v", state)
```

File: pkg/network/protocols/postgres/telemetry.go [hunk 20]
```diff
@@ -22,18 +22,96 @@ const (
...
+func newExtractionFailureCounter(metricGroup *libtelemetry.MetricGroup, metricName string, tags ...string) *extractionFailureCounter {
+	return &extractionFailureCounter{
+		countTableAndOperationFound: metricGroup.NewCounter(metricName, append(tags, "state:table_and_op")...),
+		countOperationNotFound:      metricGroup.NewCounter(metricName, append(tags, "state:no_operation")...),
+		countTableNameNotFound:      metricGroup.NewCounter(metricName, append(tags, "state:no_table_name")...),
+		countTableAndOpNotFound:     metricGroup.NewCounter(metricName, append(tags, "state:no_table_no_op")...),
+	}
+}
+
+// inc increments the counter of completed results or counter incomplete results.
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry.go:71]:
```suggestion
// inc increments the appropriate counter based on the provided state.
```

File: pkg/network/protocols/postgres/telemetry_test.go [hunk 21]
```diff
@@ -187,8 +194,17 @@ func createEbpfEvent(querySize int) *ebpf.EbpfEvent {
 }
 
 func verifyTelemetry(t *testing.T, tel *Telemetry, expected telemetryResults) {
+
```
Comment (amitslavin) [inline pkg/network/protocols/postgres/telemetry_test.go:197]: Remove the redundant new line

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/postgres/telemetry.go [hunk 22]
```diff
@@ -22,18 +22,96 @@ const (
...
+	case tableAndOperation:
+		return "tableAndOperation"
+	case operationNotFound:
+		return "operationNotFound"
+	case tableNameNotFound:
+		return "tableNameNotFound"
+	case tableAndOpNotFound:
+		return "tableAndOpNotFound"
+	default:
+		return fmt.Sprintf("%d", int(e))
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go:45]: This function should print "state" value in test_telemetry in case of failure

**Review by amitslavin — APPROVED** [review]
